### PR TITLE
pybind: add WriteOp::set_xattr() & rm_xattr()

### PR DESCRIPTION
**Pull request by zjcmszh**
Signed-off-by: Zhang Jiao <zhangjiao@cmss.chinamobile.com>

### CODE REVIEW

**Comment by zjcmszh** [comment]
Test code:
```
from rados import (Rados, WriteOpCtx, LIBRADOS_CREATE_EXCLUSIVE, NoData)
from nose.tools import eq_ as eq, ok_ as ok, assert_raises
import threading
import rados
import unittest

class TestIoctx(unittest.TestCase):
    def setUp(self):
        self.rados = Rados(conffile='./ceph.conf')
        self.rados.connect()
        self.rados.create_pool('test_pool10')
        assert self.rados.pool_exists('test_pool10')
        self.ioctx = self.rados.open_ioctx('test_pool10')
    def tearDown(self):
        self.ioctx.close()
        self.rados.delete_pool('test_pool10')
        self.rados.shutdown()

    def test_xattrs_op(self):
        with WriteOpCtx() as write_op:
            write_op.new(LIBRADOS_CREATE_EXCLUSIVE)
            write_op.setxattr('key', 'value')
            self.ioctx.operate_write_op(write_op, "abc")
            eq(self.ioctx.get_xattr('abc', 'key'), 'value')
            write_op.rmxattr('key')
            self.ioctx.operate_write_op(write_op, "abc")
            with assert_raises(NoData):
                self.ioctx.get_xattr('abc', 'key')
            write_op.remove()

if __name__ == '__main__':
    unittest.main()
```
 Test results:
```
[root@jiaowa build]# python test_xattr.py 
.
----------------------------------------------------------------------
Ran 1 test in 3.886s

OK

```

File: src/pybind/rados/rados.pyx [hunk 1]
```diff
@@ -2046,6 +2048,36 @@ cdef class WriteOp(object):
         with nogil:
             rados_write_op_set_flags(self.write_op, _flags)
 
+    @requires(('xattr_name', str_type), ('xattr_value', bytes))
+    def setxattr(self, xattr_name, xattr_value):
+        """
+        Set an xattr
+        :param xattr_name : name of the xattr
```
Comment (ideepika) [inline src/pybind/rados/rados.pyx]: can you remove the extra whitespace from parameters, something like `:param xattr_name:`

Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: @ideepika updated. please check.

Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: thanks a lot

File: src/pybind/rados/rados.pyx [hunk 2]
```diff
@@ -294,7 +294,9 @@ cdef extern from "rados/librados.h" nogil:
     void rados_write_op_omap_rm_keys(rados_write_op_t write_op, const char * const* keys, size_t keys_len)
     void rados_write_op_omap_clear(rados_write_op_t write_op)
     void rados_write_op_set_flags(rados_write_op_t write_op, int flags)
-
+    void rados_write_op_setxattr(rados_write_op_t write_op, const char *name, const char *value, size_t value_len)
```
Comment (ideepika) [inline src/pybind/rados/rados.pyx]: can you update instances of `s/setxattr/set_xattr` and `s/rmxattr/rm_xattr`

File: src/pybind/rados/rados.pyx [hunk 3]
```diff
@@ -2046,6 +2048,36 @@ cdef class WriteOp(object):
...
+        :param xattr_value: buffer to set xattr to
+        :type xattr_value: bytes
+        """
+        xattr_name = cstr(xattr_name, 'xattr_name')
+        cdef:
+            char *_xattr_name = xattr_name
+            char *_xattr_value = xattr_value
+            size_t _xattr_value_len = len(xattr_value)
+        with nogil:
+            rados_write_op_setxattr(self.write_op, _xattr_name, _xattr_value, _xattr_value_len)
```
[the remark this answers is not in the document]
Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: rados_write_op_setxattr() doesn't return value, I don't think it's necessary to catch exceptions here.

File: src/pybind/rados/rados.pyx (hunk 2 above)
Comment (zjcmszh) [reply src/pybind/rados/rados.pyx]: @ideepika updated, please review

File: src/pybind/rados/rados.pyx [hunk 4]
```diff
@@ -2046,6 +2048,36 @@ cdef class WriteOp(object):
...
+        """
+        Remove an xattr 
+        :param xattr_name: name of the xattr to remove
+        :type xattr_name: str
+        """
+        xattr_name = cstr(xattr_name, 'xattr_name')
+        cdef:
+            char *_xattr_name = xattr_name
+        with nogil:
+            rados_write_op_rmxattr(self.write_op, _xattr_name)
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/pybind/rados/rados.pyx]: please refer to the declaration of `rados_write_op_rmxattr()`
```c++
/**
 * Remove an xattr
 * @param write_op operation to add this action to
 * @param name name of the xattr to remove
 */
CEPH_RADOS_API void rados_write_op_rmxattr(rados_write_op_t write_op,
                                           const char *name);
```
this API does not return an error.

**Comment by tchaikov** [comment]
@zjcmszh i recommend you to move related changes into a single commit, i.e. to put `set_xattr()` changes to a commit, and to put `rm_xattr()` changes to another. so every commit is atomic and self-contained.

**Comment by zjcmszh** [comment]
@tchaikov    
 ```
   def test_xattrs_op(self):
        with WriteOpCtx() as write_op:
            write_op.new(LIBRADOS_CREATE_EXCLUSIVE)
            write_op.set_xattr('key', 'value')
            self.ioctx.operate_write_op(write_op, "abc")
            eq(self.ioctx.get_xattr('abc', 'key'), 'value')
            write_op.rm_xattr('key')
            self.ioctx.operate_write_op(write_op, "abc")
            with assert_raises(NoData):
                self.ioctx.get_xattr('abc', 'key')
            write_op.remove()
```
thanks a lot,  but this test method test both WriteOP::set_xattr()  and WriteOP::rm_xattr()

**Comment by zjcmszh** [comment]
updated, please review, thanks a lot.

**Comment by zjcmszh** [comment]
@tchaikov  updated, please review, thanks a lot.

**Comment by tchaikov** [comment]
jenkins test make check

File: src/test/pybind/test_rados.py [hunk 5]
```diff
@@ -583,6 +583,18 @@ def test_clear_omap(self):
             self.ioctx.operate_read_op(read_op, "hw")
             eq(list(iter), [])
 
+    def test_xattrs_op(self):
+        with WriteOpCtx() as write_op:
+            write_op.new(LIBRADOS_CREATE_EXCLUSIVE)
```
Comment (ideepika) [inline src/test/pybind/test_rados.py]: can you test multiple xattrs storing keys and values in a dictionary and hence checking multiple cases in one test?

Comment (tchaikov) [reply src/test/pybind/test_rados.py]: @zjcmszh again, please test the new test cases. you need to import this variable from rados module.

```
2019-12-09T13:16:35.865 INFO:tasks.workunit.client.0.smithi139.stderr:======================================================================
2019-12-09T13:16:35.865 INFO:tasks.workunit.client.0.smithi139.stderr:ERROR: test_rados.TestIoctx.test_xattrs_op
2019-12-09T13:16:35.865 INFO:tasks.workunit.client.0.smithi139.stderr:----------------------------------------------------------------------
2019-12-09T13:16:35.865 INFO:tasks.workunit.client.0.smithi139.stderr:Traceback (most recent call last):
2019-12-09T13:16:35.865 INFO:tasks.workunit.client.0.smithi139.stderr:  File "/usr/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
2019-12-09T13:16:35.866 INFO:tasks.workunit.client.0.smithi139.stderr:    self.test(*self.arg)
2019-12-09T13:16:35.866 INFO:tasks.workunit.client.0.smithi139.stderr:  File "/home/ubuntu/cephtest/clone.client.0/src/test/pybind/test_rados.py", line 588, in test_xattrs_op
2019-12-09T13:16:35.866 INFO:tasks.workunit.client.0.smithi139.stderr:    write_op.new(LIBRADOS_CREATE_EXCLUSIVE)
2019-12-09T13:16:35.867 INFO:tasks.workunit.client.0.smithi139.stderr:NameError: global name 'LIBRADOS_CREATE_EXCLUSIVE' is not defined
2019-12-09T13:16:35.867 INFO:tasks.workunit.client.0.smithi139.stderr:
2019-12-09T13:16:35.867 INFO:tasks.workunit.client.0.smithi139.stderr:----------------------------------------------------------------------
```

Comment (zjcmszh) [reply src/test/pybind/test_rados.py]: 我立马修改,稍等

Just a moment,  I'll fix it right away.

Comment (zjcmszh) [reply src/test/pybind/test_rados.py]: Just a moment , I'll fix it right away.

Comment (zjcmszh) [reply src/test/pybind/test_rados.py]: test code:
```
#from rados import (Rados, WriteOpCtx, LIBRADOS_CREATE_EXCLUSIVE, NoData)
from rados import (Rados, Error, RadosStateError, Object, ObjectExists,
                   ObjectNotFound, ObjectBusy, requires, opt,
                   LIBRADOS_ALL_NSPACES, WriteOpCtx, ReadOpCtx, LIBRADOS_CREATE_EXCLUSIVE,
                   LIBRADOS_SNAP_HEAD, LIBRADOS_OPERATION_BALANCE_READS, LIBRADOS_OPERATION_SKIPRWLOCKS, MonitorLog, NoData)

from nose.tools import eq_ as eq, ok_ as ok, assert_raises
import threading
import rados
import unittest

class TestIoctx(unittest.TestCase):
    def setUp(self):
        self.rados = Rados(conffile='./ceph.conf')
        self.rados.connect()
        self.rados.create_pool('poll10')
        assert self.rados.pool_exists('poll10')
        self.ioctx = self.rados.open_ioctx('poll10')
    def tearDown(self):
        self.ioctx.close()
        self.rados.delete_pool('poll10')
        self.rados.shutdown()

    def test_xattrs_op(self):
        with WriteOpCtx() as write_op:
            write_op.new(LIBRADOS_CREATE_EXCLUSIVE)
            write_op.set_xattr('key', 'value')
            self.ioctx.operate_write_op(write_op, "abc")
            eq(self.ioctx.get_xattr('abc', 'key'), 'value')
            write_op.rm_xattr('key')
            self.ioctx.operate_write_op(write_op, "abc")
            with assert_raises(NoData):
                self.ioctx.get_xattr('abc', 'key')
            write_op.remove()

if __name__ == '__main__':
    unittest.main()

```
test result:
```
[root@jiaowa build]# python test_xattr.py
.
----------------------------------------------------------------------
Ran 1 test in 3.665s

OK
```
@tchaikov  updated, please check, very very thankful

Comment (zjcmszh) [reply src/test/pybind/test_rados.py]: thanks ，I'll fix it now.

**Comment by zjcmszh** [comment]
test code:
```
from rados import (Rados, WriteOpCtx, LIBRADOS_CREATE_EXCLUSIVE)
from nose.tools import eq_ as eq, ok_ as ok, assert_raises
import threading
import rados
import unittest

class TestIoctx(unittest.TestCase):
    def setUp(self):
        self.rados = Rados(conffile='./ceph.conf')
        self.rados.connect()
        self.rados.create_pool('poll10')
        assert self.rados.pool_exists('poll10')
        self.ioctx = self.rados.open_ioctx('poll10')
    def tearDown(self):
        self.ioctx.close()
        self.rados.delete_pool('poll10')
        self.rados.shutdown()

    def test_xattrs_op(self):
        xattrs = dict(a=b'1', b=b'2', c=b'3', d=b'a\0b', e=b'\0')
        with WriteOpCtx() as write_op:
            write_op.new(LIBRADOS_CREATE_EXCLUSIVE)
            for key, value in xattrs.items():
                write_op.setxattr(key, value)
                self.ioctx.operate_write_op(write_op, "abc")
                eq(self.ioctx.get_xattr('abc', key), value)
            stored_xattrs_1 = {}
            for key, value in self.ioctx.get_xattrs('abc'):
                stored_xattrs_1[key] = value
            eq(stored_xattrs_1, xattrs)
            for key in xattrs.keys():
                write_op.rmxattr(key)
                self.ioctx.operate_write_op(write_op, "abc")
            stored_xattrs_2 = {}
            for key, value in self.ioctx.get_xattrs('abc'):
                stored_xattrs_2[key] = value
            eq(stored_xattrs_2, {})
            write_op.remove()

if __name__ == '__main__':
    unittest.main()

```
test results:
```
[root@jiaowa build]# python test_xattr.py
.
----------------------------------------------------------------------
Ran 1 test in 2.837s

OK
```
@ideepika  udpated, please check

File: src/test/pybind/test_rados.py (hunk 5 above)
Comment (zjcmszh) [reply src/test/pybind/test_rados.py]: updated, please check.

**Comment by zjcmszh** [comment]
@ideepika  if there is no problem, could you approve it ? 
thanks a lot

**Comment by zjcmszh** [comment]
@ideepika If there are no problems, would you approve it? vary vary vary very thankful

File: src/test/pybind/test_rados.py [hunk 6]
```diff
@@ -583,6 +583,27 @@ def test_clear_omap(self):
             self.ioctx.operate_read_op(read_op, "hw")
             eq(list(iter), [])
 
+    def test_xattrs_op(self):
+        xattrs = dict(a=b'1', b=b'2', c=b'3', d=b'a\0b', e=b'\0')
+        with WriteOpCtx() as write_op:
+            write_op.new(LIBRADOS_CREATE_EXCLUSIVE)
+            for key, value in xattrs.items():
+                write_op.setxattr(key, value)
```
Comment (ideepika) [inline src/test/pybind/test_rados.py]: think you are not testing set_xattr method, can you update and paste test run?

Comment (zjcmszh) [reply src/test/pybind/test_rados.py]: @ideepika udpated,please check. thanks.
test code:
```
from rados import (Rados, WriteOpCtx, LIBRADOS_CREATE_EXCLUSIVE)
from nose.tools import eq_ as eq, ok_ as ok, assert_raises
import threading
import rados
import unittest

class TestIoctx(unittest.TestCase):
    def setUp(self):
        self.rados = Rados(conffile='./ceph.conf')
        self.rados.connect()
        self.rados.create_pool('poll10')
        assert self.rados.pool_exists('poll10')
        self.ioctx = self.rados.open_ioctx('poll10')
    def tearDown(self):
        self.ioctx.close()
        self.rados.delete_pool('poll10')
        self.rados.shutdown()

    def test_xattrs_op(self):
        xattrs = dict(a=b'1', b=b'2', c=b'3', d=b'a\0b', e=b'\0')
        with WriteOpCtx() as write_op:
            write_op.new(LIBRADOS_CREATE_EXCLUSIVE)
            for key, value in xattrs.items():
                write_op.set_xattr(key, value)
                self.ioctx.operate_write_op(write_op, "abc")
                eq(self.ioctx.get_xattr('abc', key), value)
            stored_xattrs_1 = {}
            for key, value in self.ioctx.get_xattrs('abc'):
                stored_xattrs_1[key] = value
            eq(stored_xattrs_1, xattrs)
            for key in xattrs.keys():
                write_op.rm_xattr(key)
                self.ioctx.operate_write_op(write_op, "abc")
            stored_xattrs_2 = {}
            for key, value in self.ioctx.get_xattrs('abc'):
                stored_xattrs_2[key] = value
            eq(stored_xattrs_2, {})
            write_op.remove()

if __name__ == '__main__':
    unittest.main()
```
test result:
```
[root@jiaowa build]# python test_xattr.py
.
----------------------------------------------------------------------
Ran 1 test in 2.316s

OK
```

**Comment by ideepika** [comment]
Jenkins make check, make check arm64 please

**Comment by ideepika** [comment]
jenkins retest this please

**Comment by ideepika** [comment]
please update commit message for 372bb6e2d34d38084e4370b49137454a4848a577 to include tests as well for set_xattr function

**Comment by zjcmszh** [comment]
```
    def test_xattrs_op(self):
        xattrs = dict(a=b'1', b=b'2', c=b'3', d=b'a\0b', e=b'\0')
        with WriteOpCtx() as write_op:
            write_op.new(LIBRADOS_CREATE_EXCLUSIVE)
            for key, value in xattrs.items():
                write_op.set_xattr(key, value)     #test set_xattr()
                self.ioctx.operate_write_op(write_op, "abc")
                eq(self.ioctx.get_xattr('abc', key), value)
            stored_xattrs_1 = {}
            for key, value in self.ioctx.get_xattrs('abc'):
                stored_xattrs_1[key] = value
            eq(stored_xattrs_1, xattrs)
            for key in xattrs.keys():
                write_op.rm_xattr(key)
                self.ioctx.operate_write_op(write_op, "abc")
            stored_xattrs_2 = {}
            for key, value in self.ioctx.get_xattrs('abc'):
                stored_xattrs_2[key] = value
            eq(stored_xattrs_2, {})
            write_op.remove()
```
@ideepika  set _xattr() has been tested.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-12-26_05:49:41-rados-wip-kefu-testing-2019-12-26-1055-distro-basic-smithi/

failures tracked / fixed by

- https://tracker.ceph.com/issues/38539
- https://tracker.ceph.com/issues/43396
- https://github.com/ceph/ceph/pull/32427
- https://github.com/ceph/ceph/pull/32426
- https://github.com/ceph/ceph/pull/32428
- #32429

**Comment by ideepika** [comment]
jenkins test make check

**Pull request merged**